Thesis.org

```diff
@@ -1,11 +1,725 @@
+See listing 28 to see how the nesting of anonymous functions works out, to allow ~end-expr~ to have access to the symbols at the end.
+#+CAPTION: Example of where ~end-expr~ will be executed.
+#+begin_src clojure
+  (then prog1
+	(fn [x]
+	  (then prog2
+		(fn [y]
+		  (then prog3
+			;; assuming end-expr accesses x y z symbols
+			(fn [z] (end-expr)))))))
+#+end_src
+Now our ~then~ macro can be called exactly like we would with ~let~, simply by doing the following:  ~(then [a prog1 b prog2] (fn [] (+ a b)))~ we are able to chain together ~prog1~ and ~prog2~. What we get back is a ~Bind~ of both ~Progs~.
+*** runner
+With our ~then~ we have greatly improved the desired developer experience, however, the result still returns a ~Bind~. That is fine, as we might want to further compose this. However, we still need to wrap our ~then~ expression with a ~runner~ to run it. To further simplify the experience we will create another macro, this time for the ~runner~ function to give developers the option to do everything within a single call.
+For this we will simply wrap our ~then~ macro with our primitive ~runner~ from the ~bind~ namespace. Again we syntax quote our call and undo the quote for our values. To add even more convenience our ~runner~ macro has an arity of two. If the developer is not interested in supplying an ~end-expr~ function, we will pass an empty function into the ~then~ for them. The resulting code is simple but works exactly as we want it to and can be seen in listing 29.
+#+CAPTION: Definition of the ~runner~ macro.
+#+begin_src clojure
+  (defmacro runner
+    ([x]
+     `(runner ~x (fn [])))
+    ([x y]
+     `(code.bind/runner (then ~x  ~y))))
+#+end_src
+** Limitations
+While the current implementation achieves what it set out to do, some compromises had to be made. First, as mentioned earlier, it is being relied on the fact that the developer actually emits a ~Commit~ in what they label a ~Prog~. There is currently no logic to make sure that the developer is forced too or reminded if they aren't. Another limitation is that the current API offers only primitives and our two macros, but none of the deep functionality which is found in reacl-c for the parallel composability. So functions that map, filter etc. over sequential compositions are not included. There is also no error handling for sequential composition. Developers need to handle errors by hand in the continuation of the next ~Bind~, as there is no ~Error~ sub-type of ~Commit~. Lastly, there is no way to terminate early, like with a ~Maybe~ monad.
```

Comment:
> it is being relied on the fact that the developer actually emits a ~Commit~ in what they label a ~Prog~

A nice way to solve this is to offer only a primitive `wait-for-action : Item -> Predicate -> Prog`. `wait-for-action` takes an item and listens for actions. When an action satisfies `Predicate` it returns a `Commit`. The `Commit` record itself could then be an internal implementation detail of the library.

Reply:
> So functions that map, filter etc. over sequential compositions are not included.

These could simply be written via `bind` and `return`. 

In Haskell: `fmap f x = x >>= (return . f)`

Reply:
I like the idea of ``wait-for-action`` and I agree that ``fmap`` would be trivial. I see both as a v1.1 kinda deal, therefor mention it under "Limitations".